### DISCUSSION TITLE
Plugin settings

### DISCUSSION THREAD

**Discussion by rolandlo** [xournalpp/xournalpp#6060] [Ideas]
Xournal++ plugins may want to save some settings. We currently don't have a standardized way to do that. There should be at least a guide how to do it. One problem we face is that plugins shipped with Xournal++ are installed in a system folder (like the folder `/usr/share/xournalpp/plugins/<PluginName>` on Linux), where only an admin can write. Here is my suggestion:

- plugin settings should go into  `XDG_CONFIG_HOME/xournalpp/plugin-settings/<PluginName>`
- to get the location of the Plugin's settings folder (in a cross platform way), the Lua API should be extended say with `app.getPluginSettingsFolder()` or maybe in a more extendable way by `app.getFolder("settings")`
- maybe add an option `force` which lets you chose whether the folder should be created if it doesn't exist already (so it's `app.getPluginSettingsFolder(force)` or `app.getFolder("settings", force)` with a boolean named `force` that is true by default.
- we may also add `app.getFolder("data", force)` to return a plugin data folder `XDG_DATA_HOME/xournalpp/plugin-data/<PluginName>`
- Two plugins with the same name can be loaded (like plugin `Foo` from `/usr/share/xournalpp/plugins/Foo` and one from `~/.config/xournalpp/plugins/Foo`), which complicates matters. I would suggest they use the same settings and data (folder).

Loosely related to that, user plugins should (at least additionally) be loaded from `XDG_DATA_HOME` (currently only `XDG_CONFIG_HOME`) as suggested in https://github.com/xournalpp/xournalpp/issues/4657#issuecomment-1520411101.

@bhennion @Febbe @atticus-sullivan @MiltonBalaOfficial what's your take on that?

**Comment by atticus-sullivan** [comment]
Yes sounds like a good idea. If we really want to, we could try to add measures to further distinguish plugins with the same name, but I'm not sure if this really is needed.

It might be worth thinking about whether we can/want to somehow ship something for parsing config files in order to avoid each plugin shipping its own parsing lib (of course they could rely on a lib being installed like with `lgi` (e.g. via LuaRocks) but I feel like this always makes using plugins so much harder if you're not too familiar with lua). But probably this is out of scope for the plugin api. Or at least it's a different discussion.

**Reply by rolandlo** [reply]
I think the usual way of using config files in Lua (even by projects like luarocks or neovim) is to just make them be Lua files and executing the file instead of parsing its content. The config file can also return a table that is obtained via `require` or `dofile`. See https://github.com/xournalpp/xournalpp/pull/6020 for an example where the configuration is obtained via `shapes_dict = require "config"`.

**Reply by atticus-sullivan** [reply]
True, lua script as config file is easy to parse. Writing/Generating it can be a bit tedious (there is the `%q` format specifier but as usual in lua, this does not help with complex datatypes like tables/arrays).

Still as already outlined, packaging a complete xml/json/yaml parsing lib just for that one plugin which wants to do fancy config file parsing is probably a bit overkill.

**Comment by rolandlo** [comment]
See #6155 for a PR.
